### PR TITLE
Makefile Shared Library

### PR DESCRIPTION
**Pull request by mattyclarkson**
This adds shared library support in the makefile

### CODE REVIEW

**Comment by pau-climent** [comment]
I've just applied for/signed a CLA. How can this now be solved?

File: Makefile [hunk 1]
```diff
@@ -69,12 +70,12 @@ $(CCTZ_LIB): $(CCTZ_OBJS)
 install: install_hdrs install_lib
 
 install_hdrs: $(CCTZ_HDRS)
-	$(SUDO) mkdir -p $(PREFIX)/include
-	$(SUDO) cp -p $? $(PREFIX)/include
+	$(SUDO) mkdir -p $(DESTDIR)/$(PREFIX)/include
```
Comment (devbww) [inline Makefile]: What is it that DESTDIR is buying us over PREFIX?  They always appear together.  Should we just rename PREFIX to DESTDIR?

File: Makefile [hunk 2]
```diff
@@ -38,10 +38,17 @@ CPPFLAGS = -Wall -I$(SRC)include -std=$(STD) -pthread \
...
+LN = ln
+CP = cp
+MKDIR = mkdir
 
 SUDO =
 
 CCTZ_LIB = libcctz.a
+CCTZ_STATIC_LIB = $(CCTZ_LIB)
+CCTZ_SHARED_LIB = libcctz.so.2.0
+CCTZ_SHARED_SONAME = $(basename $(CCTZ_SHARED_LIB))
```
Comment (devbww) [inline Makefile]: Why use basename on paths that do not contain slashes?

File: Makefile [hunk 3]
```diff
@@ -62,25 +69,35 @@ EXAMPLES = classic epoch_shift hello example1 example2 example3 example4
 
 all: $(TESTS) $(TOOLS) $(EXAMPLES)
 
-$(TESTS) $(TOOLS) $(EXAMPLES): $(CCTZ_LIB)
+$(TESTS) $(TOOLS) $(EXAMPLES): $(CCTZ_SHARED_LIB)
 
-$(CCTZ_LIB): $(CCTZ_OBJS)
+$(CCTZ_SHARED_LIB): $(CCTZ_OBJS) $(CCTZ_SHARED_SONAME)
```
Comment (devbww) [inline Makefile]: There seems to be some circularity here ... CCTZ_SHARED_LIB depends on CCTZ_SHARED_SONAME and CCTZ_SHARED_SONAME depends on CCTZ_SHARED_LIB.

File: Makefile [hunk 4]
```diff
@@ -38,10 +38,17 @@ CPPFLAGS = -Wall -I$(SRC)include -std=$(STD) -pthread \
...
 LDLIBS = $(TEST_LIBS)
+LN = ln
+CP = cp
+MKDIR = mkdir
 
 SUDO =
 
 CCTZ_LIB = libcctz.a
+CCTZ_STATIC_LIB = $(CCTZ_LIB)
+CCTZ_SHARED_LIB = libcctz.so.2.0
```
Comment (devbww) [inline Makefile]: Is "2.0" referring to "CCTZ Version 2"?  If so, perhaps we should be providing some more central versioning information.

File: Makefile [hunk 5]
```diff
@@ -42,6 +42,7 @@ LDLIBS = $(TEST_LIBS)
 SUDO =
 
 CCTZ_LIB = libcctz.a
+CCTZ_STATIC_LIB = $(CCTZ_LIB)
```
Comment (mattyclarkson) [inline Makefile]: Yeah, makes space for `CCTZ_SHARED_LIB`. The `CCTZ_LIB` was just left there as backwards compatible variable in case anyone was doing `make CCTZ_LIB=libmycctz.a` for some reason. Happy to remove it!

File: Makefile (hunk 2 above)
Comment (mattyclarkson) [inline Makefile]: It removes the last part of the version number, so it becomes `libcctz.so.2` (drops the `.0`)

File: Makefile (hunk 4 above)
Comment (mattyclarkson) [inline Makefile]: Sure, usually a `VERSION` file is used. Would happily read that file to determine the version. I initially did a describe of the lastest `git tag` but realised that it wouldn't work for archives.

File: Makefile (hunk 1 above)
Comment (mattyclarkson) [inline Makefile]: `DESTDIR` is a common override that is used in `autotools` style makefiles. See the [documentation](https://www.gnu.org/software/automake/manual/html_node/DESTDIR.html).

The reason to have a seperate `PREFIX` and `DESTDIR` is that the prefix is the installation prefix but the DESTDIR allows installation into a different location. This is especially useful when cross-compiling as the `PREFIX` is actually `/usr` but the `DESTDIR` is something like `/home/user/cc-staging`. Then the `DESTDIR` is packaged up and expanded onto the target device.

Some projects use the `prefix` variable to understand where the project is to be installed eventually. This project doesn't use that at all in the C++ code but it would be nice if we can play nice with common `Makefile` conventions. Happy to remove it if you don't want it though :smile:

File: Makefile (hunk 5 above)
Comment (devbww) [inline Makefile]: Let's just leave CCTZ_LIB as it is then.  That is, forget about CCTZ_STATIC_LIB ... just introduce CCTZ_SHARED_LIB.

File: Makefile (hunk 4 above)
Comment (devbww) [inline Makefile]: OK, thanks.  Let's just leave it as a literal "2.0" for now, then.

File: Makefile (hunk 1 above)
Comment (devbww) [inline Makefile]: OK, then let's at least remove the /'s between $(DESTDIR) and $(PREFIX).

File: Makefile (hunk 3 above)
Comment (mattyclarkson) [inline Makefile]: Whoops, sorry that's not a dependency.

**Comment by mattyclarkson** [comment]
@devbww should be good to go unless you have extra comments.

**Comment by mattyclarkson** [comment]
Once this goes in; are there any plans to do another release? It would be really helpful to have an archive that we can download and integrate into our distribution that can build shared libraries directly from the makefile. We currently are running with patches on top of the v2.0 tarball.

File: Makefile [hunk 6]
```diff
@@ -69,6 +73,15 @@ $(TESTS) $(TOOLS) $(EXAMPLES): $(CCTZ_LIB)
 
 $(TESTS:=.o) $(TOOLS:=.o) $(EXAMPLES:=.o):
 
+$(CCTZ_SHARED_LIB): $(CCTZ_OBJS)
+	$(CC) $(LDFLAGS) -shared -Wl,-soname,$(CCTZ_SHARED_SONAME) -o$@ $(CCTZ_OBJS)
+
+$(CCTZ_SHARED_SONAME): $(CCTZ_SHARED_LIB)
```
Comment (devbww) [inline Makefile]: Is it important for these links to be in the build tree, or are they best handled by the installation rule?

File: Makefile [hunk 7]
```diff
@@ -67,7 +67,7 @@ CCTZ_OBJS =			\
 TOOLS = time_tool
 EXAMPLES = classic epoch_shift hello example1 example2 example3 example4
 
-all: $(TESTS) $(TOOLS) $(EXAMPLES)
+all: $(TESTS) $(TOOLS) $(EXAMPLES) $(CCTZ_SHARED_LINK)
```
Comment (devbww) [inline Makefile]: Why is this required?

File: Makefile [hunk 8]
```diff
@@ -39,12 +39,16 @@ ARFLAGS = rcs
...
 LDLIBS = $(TEST_LIBS)
 
+LN = ln
 CP = cp
 MKDIR = mkdir
 SUDO =
 
 CCTZ = cctz
 CCTZ_LIB = lib$(CCTZ).a
+CCTZ_SHARED_LIB = lib$(CCTZ).so.2.0
```
Comment (devbww) [inline Makefile]: Perhaps the GNU dependency is already there, with VPATH and -include, but perhaps it is still clearer this way.

**Comment by devbww** [comment]
> Once this goes in; are there any plans to do another release? It would be really helpful to have an archive that we can download and integrate into our distribution that can build shared libraries directly from the makefile. We currently are running with patches on top of the v2.0 tarball.

Yes, we can do that after a few other pending changes land.  Thanks.

File: Makefile (hunk 7 above)
Comment (mattyclarkson) [inline Makefile]: It's not. Was just so that the shared library ended up in the build folder when doing `make all`. Will remove it.

File: Makefile (hunk 6 above)
Comment (mattyclarkson) [inline Makefile]: Nope. They can be done by the install. Will do that.

File: Makefile (hunk 8 above)
Comment (mattyclarkson) [inline Makefile]: I added a `CCTZ_VERSION_MAJOR` and `CCTZ_VERSION_MINOR` variable to make it a bit more obvious what those numbers are. Also allows those two variables to be overridden.

File: Makefile (hunk 6 above)
Comment (mattyclarkson) [inline Makefile]: Hmm, having these rules makes it a bit easier to build the library as you can do `make libcctz.so` rather than having to know the version number (`make libcctz.so.2.0`). I've left them in but I can remove them if you don't see that as a benefit.

File: Makefile [hunk 9]
```diff
@@ -69,6 +73,15 @@ $(TESTS) $(TOOLS) $(EXAMPLES): $(CCTZ_LIB)
 
 $(TESTS:=.o) $(TOOLS:=.o) $(EXAMPLES:=.o):
 
+$(CCTZ_SHARED_LIB): $(CCTZ_OBJS)
+	$(CC) $(LDFLAGS) -shared -Wl,-soname,$(CCTZ_SHARED_SONAME) -o$@ $(CCTZ_OBJS)
```
Comment (mattyclarkson) [inline Makefile]: Done.
